src/OpenTelemetry/Trace/Export/SimpleActivityProcessor.cs

```diff
@@ -46,17 +105,16 @@ public override void OnStart(Activity activity)
         /// <inheritdoc />
         public override void OnEnd(Activity activity)
         {
-            try
+            if (this.currentQueueSize >= this.maxQueueSize)
             {
-                // do not await, just start export
-                // it can still throw in synchronous part
-
-                _ = this.exporter.ExportAsync(new[] { activity }, CancellationToken.None);
-            }
-            catch (Exception ex)
-            {
-                OpenTelemetrySdkEventSource.Log.SpanProcessorException(nameof(this.OnEnd), ex);
+                OpenTelemetrySdkEventSource.Log.SpanProcessorQueueIsExhausted();
+                return;
             }
+
+            Interlocked.Increment(ref this.currentQueueSize);
+
+            this.activityQueue.Enqueue(activity);
```

Comment:
I'd let @reyang comment here, but introducing a queue for simpleprocessor defeats the purpose it was intended, won't it?

Reply:
Yeah, I was bringing this topic to the [OpenTelemetry Specification SIG meeting 07/14/2020](https://docs.google.com/document/d/1-bCYkN-DWJq4jw1ybaDZYYmx-WAe6HnwfWbkm8d57v8/edit#heading=h.p9hvwjnz3snv) as topic 8.

Have seen the challenge on .NET, C++ and other scenarios/languages that require high performance / concurrency. For example, an exporter that concurrently writes data to shared memory, ETW (event tracing for Windows) or LTTng.

This is something I need to work on from the spec perspective.

Reply:
Open to ideas, but it feels unavoidable. How we had it before, starting Tasks, that's just queuing work for the thread pool, right?

Reply:
My current thinking:
1. simple processor should run concurrently without contention.
2. exporter could have synchronization by default, following the spec (unless I changed the spec before GA).
3. we probably need a way for certain exporter to express that they are "thread free", similar like the COM STA/MTA model, so that the SDK won't try to synchronize the call.

Reply:
I thought this one might be tricky. I agree with you guys, 100%, but the current version I don't think is passable. It is going to flood the thread pool under load. Can we move forward with this more safe design and pursue clarity with the spec for GA? Keep in mind, the simple processor is currently the default.

If an exporter is fast enough, this will export spans one-by-one as they are ready. Tight inner loop. It is only when the exporter is slow that we start feeding it chunks of data. Once our queue fills up, we start dropping data. It's a more safe approach all around IMO.

Satisfies the spec mandate that we shouldn't call export concurrently, but also more important mandates like we shouldn't starve the hosting process 🍽️

Reply:
I'm worried that we move forward with this direction.
The Python SDK is actually using [lock inside exporter](https://github.com/open-telemetry/opentelemetry-python/blob/5ff9600a62f9c82fc120b03e5729a53341465626/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/in_memory_span_exporter.py#L49), which means the exporter code is called concurrently.
The C++ SDK is also hitting the same concern and that's why I raised the question in the specification SIG.

Give me some time to do a quick experiment and I might come up with some solution.

Reply:
Sure, looking forward to seeing what you come up with!

I kind of dig option 4. We provide a couple of OOB solutions geared towards best-effort, safety, and low-impact on the hosting process but if you want to make something ultra performant you can cut out the exporting layer completely and do it directly in the processor.

3 we should probably do regardless? At least get clarification on how it should work.

1 I'm skeptical that any kind of synchronization will be successful. I'm imagining a busy process creating a lof of spans very quickly with a slow exporter.

2 I thought about changing the interface when I was doing the work but it didn't really help the situation. The way it is written today (async) with the fire-and-forget we'll flood the pool. If it was sync, we'd block the SpanProcessor.EndActivity from finishing which holds up that thread. Neither case is really ideal.

Reply:
I like this design and maybe we can port some of the work to the batching exporter. Using the handles to stop the thread constantly checking for work is nice.

I also agree with @cijothomas and @reyang that the simple processor is specced to not batch spans, the simple span exporter is effectively designed to be a simple queue that can be exhausted fairly easily. It is strange that the default will naturally lack performance and we should pursue amending the spec to default to a batching style exporter.

Reply:
@reyang Ping :)

The main thing I want to accomplish here is removing the fire-and-forget task/thread pool thing. That is soooo dangerous! In a high-volume/slow exporter situation, I'm pretty sure that will crash the process. If we want to remove the batching in the worker thread, no problem. It seems like a crime to not batch up the data when we know it's sitting there, but 🤷

Reply:
@CodeBlanch this has a dependency on the refactor work, I think we should be ready to solve it on Wed.